### PR TITLE
Making regression3 pass

### PR DESCRIPTION
**Pull request by dhermes**
- Uses latest release of `oauth2client` which incorporates google/oauth2client#126
- Pulling from `HEAD` in `httplib2` until jcgregorio/httplib2#291 and jcgregorio/httplib2#296 are incorporated
- Updating bytes/unicode handling, mostly as in #724 
- Updating `storage.batch._unpack_batch_response` to straddle Py2/Py3. The email native `str` dependency is tricky there.
- Added some more tests to cover unicode/bytes branching in library code
- Updated regression for storage to explicitly use binary when `str` is insufficient in Py3

**NOTE**: As in google/oauth2client#126, using `%s` for a Py2 and Py3 codebase caused us issues in `_unpack_batch_response`

### CODE REVIEW

**Comment by dhermes** [comment]
@tseaver PTAL

File: gcloud/storage/batch.py [hunk 1]
```diff
@@ -170,10 +170,22 @@ def __exit__(self, exc_type, exc_val, exc_tb):
 def _unpack_batch_response(response, content):
     """Convert response, content -> [(status, reason, payload)]."""
     parser = Parser()
-    faux_message = ('Content-Type: %s\nMIME-Version: 1.0\n\n%s' %
-                    (response['content-type'], content))
-
-    message = parser.parsestr(faux_message)
+    if not isinstance(content, six.binary_type):
```
Comment (tseaver) [inline gcloud/storage/batch.py]: Coercing to bytes seems to be awkward:  we want `insinstance(faux_message, str)` to be true, right?

Couldn't we write a helper function (e.g., `to_native_str()`), which would Do the Right Thing(TM) here without this dance?

File: gcloud/storage/test_batch.py [hunk 2]
```diff
@@ -347,6 +347,34 @@ def test_as_context_mgr_w_error(self):
...
+        self.assertEqual(len(result), 3)
+        self.assertEqual(result[0], ('200', 'OK', {u'bar': 2, u'foo': 1}))
+        self.assertEqual(result[1], ('200', 'OK', {u'foo': 1, u'bar': 3}))
+        self.assertEqual(result[2], ('204', 'No Content', ''))
+
+    def test_unicode(self):
+        import six
+        RESPONSE = {'content-type': u'multipart/mixed; boundary="DEADBEEF="'}
+        CONTENT = _THREE_PART_MIME_RESPONSE
+        if isinstance(CONTENT, six.binary_type):  # pragma: NO COVER  Python3
```
Comment (tseaver) [inline gcloud/storage/test_batch.py]: That `#pragma` smells wrong:  `_THREE_PART_MIME_RESPONSE` is a native string, and hence will be bytes on Python2.  Maybe we should force it to be bytes on line 378?

File: gcloud/test_connection.py [hunk 3]
```diff
@@ -330,10 +330,35 @@ def test_api_request_w_500(self):
...
-            '{}',
+            b'{}',
         )
         self.assertRaises(InternalServerError, conn.api_request, 'GET', '/')
 
+    def test_api_request_non_binary_response(self):
+        conn = self._makeMockOne()
+        http = conn._http = _Http(
+            {'status': '200', 'content-type': 'application/json'},
+            u'{}',
```
Comment (tseaver) [inline gcloud/test_connection.py]: How is this ever going to happen?  Doesn't `httplib2` always return bytes?

File: gcloud/test_credentials.py [hunk 4]
```diff
@@ -176,11 +176,13 @@ def _get_pem_key(credentials):
         SIGNATURE_STRING = 'dummy_signature'
         with _Monkey(MUT, RSA=rsa, PKCS1_v1_5=pkcs_v1_5,
                      SHA256=sha256, _get_pem_key=_get_pem_key):
-            self.assertRaises(NameError, self._callFUT,
+            self.assertRaises(UnboundLocalError, self._callFUT,
```
Comment (tseaver) [inline gcloud/test_credentials.py]: How does this change relate to the PR?

File: tox.ini [hunk 5]
```diff
@@ -74,7 +74,8 @@ commands =
     {toxinidir}/scripts/run_regression.sh
 deps =
     unittest2
-#   Use a development checkout of oauth2client until a release is made
-#   which fixes https://github.com/google/oauth2client/issues/125
-    -egit+https://github.com/google/oauth2client.git#egg=oauth2client
```
Comment (tseaver) [inline tox.ini]: Don't we need to call out `oauth2client` as an explicit dependency (pulled from PyPI, rather than checked out)?

File: gcloud/storage/test_batch.py (hunk 2 above)
Comment (dhermes) [inline gcloud/storage/test_batch.py]: The `#pragma` is for the branch miss when `tox -e cover` is run (only Py 2.7).

File: gcloud/test_connection.py (hunk 3 above)
Comment (dhermes) [inline gcloud/test_connection.py]: It is an artifact of #724. In that PR, an exception was thrown when non-bytes was returned. I handled it just by using `.encode`.

File: tox.ini (hunk 5 above)
Comment (dhermes) [inline tox.ini]: It's in `setup.py`. Maybe I'm not understanding the question.

File: gcloud/storage/batch.py (hunk 1 above)
Comment (dhermes) [inline gcloud/storage/batch.py]: Yeah that would work too. I was just trying to fix the `%s` bug with `bytes` in Py3 and didn't focus too much on the fix.

On the flip side, using `str` scares me. I'd rather be explicit. (Isn't this a Py2/Py3 best practice anway? e.g. Brett Cannon's tool complains about not using `b''` and `u''`.)

File: gcloud/test_credentials.py (hunk 4 above)
Comment (dhermes) [inline gcloud/test_credentials.py]: It doesn't. I noticed it at some point throughout and decided to fix. (`UnboundLocalError` is a subclass of `NameError`)

File: gcloud/storage/test_batch.py (hunk 2 above)
Comment (tseaver) [inline gcloud/storage/test_batch.py]: OK.  If we made `_THREE_PART_MIME_RESPONSE` a bytes literal, we could just decode it unilaterally here.

File: gcloud/test_connection.py (hunk 3 above)
Comment (tseaver) [inline gcloud/test_connection.py]: But why are we mocking a non-bytes return?  The point of #724 is that htttplib2 always gives us bytes as the payload, never unicode.

File: tox.ini (hunk 5 above)
Comment (tseaver) [inline tox.ini]: At least in the past, tox has been known to install the package-under-test without all its declared dependencies, unless they are declared explicitly in `tox.ini`.  If `tox --recreate` doesn't fail with your change, then I guess we're OK here.

File: gcloud/test_connection.py (hunk 3 above)
Comment (dhermes) [inline gcloud/test_connection.py]: I just wanted to preserve the majority of the code in #724.

Removing it leads to a branch miss in `Connection.api_request`:

```python
            if isinstance(content, six.binary_type): 
                content = content.decode('utf-8') 
            return json.loads(content)
```

I thought this approach was friendlier than raising if the response was unicode.

File: tox.ini (hunk 5 above)
Comment (dhermes) [inline tox.ini]: Yes I recreated and have no issues. Travis also creates a new tox env every time.

File: gcloud/storage/test_batch.py (hunk 2 above)
Comment (dhermes) [inline gcloud/storage/test_batch.py]: Updated and pushed over/into the last commit in this PR.

File: gcloud/storage/batch.py (hunk 1 above)
Comment (tseaver) [inline gcloud/storage/batch.py]: We likely need a comment here explaining that we have to pass native `str` to `email.parser.parsestr` (so we aren't scratching our heads about this dance in a year).

**Comment by tseaver** [comment]
LGTM, module adding the comment to `_unpack_batch_response`.

File: gcloud/storage/batch.py (hunk 1 above)
Comment (dhermes) [inline gcloud/storage/batch.py]: Well the reason we use bytes is so they can be `b''` joined; it fixed the `%s` issue with bytes in Py3.

The native `str` for email part happens below at `six.PY2`.

I can add a comment here explaining we coerce to bytes so the concat is valid in Py3. WDYT?

Comment (tseaver) [inline gcloud/storage/batch.py]: SGTM.

**Comment by dhermes** [comment]
@tseaver I made a new commit from my phone. PTAL and then I'll merge if Travis passes.

**Comment by tseaver** [comment]
LGTM

**Comment by dhermes** [comment]
@tseaver should we add `regression3` to `.travis.yaml` now that it passes?

**Comment by tseaver** [comment]
SGTM
